### PR TITLE
crimson/os/seastore: enable metadata (*_INTERNAL and *_LEAF ) allocation using OOL path

### PR DESCRIPTION
**Pull request by myoungwon**
Signed-off-by: Myoungwon Oh <myoungwon.oh@samsung.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by myoungwon** [comment]
@cyx1231st I'm prototyping this work. Before continuing further changes, I'd like to discuss two items as follows:

1. Overwrite dirty extents: current trim_dirty() creates a new extent, and rewrite it to the journal at new location. On the other hand, RBM can overwrite delta to the original extent located at RBM, while deleting old one; this means that we don't need to create a new extent, just overwriting an as-is extent to the original extent is enough, while increasing the version number. But, when the version number is reset is still a question---I think overwrite without a new extent is efficient, but need a consensus. 

2. Randomblock allocation: Due to random block allocation, a same address can be allocated and deallocated. This causes inconsistency when two metadata (LADDR_*, BACKREF_*) are allocated to RBM. Please see the below.

Normal transaction A: Address 1 (LADDR_LEAF extent) is allocated -> 1 is deallocated; so, address 1 is free by mark_space_free().
Background transaction B: ------------------------------------------------->trim_alloc is triggered after 1 is deallocated, then reuses address 1 by BACKREF_LEAF due to duplication.

In terms of transaction B, after reusing address 1 due to duplication_for_write(), address 1 is replayed by new_mapping(). At this point, if trim_alloc is stopped, BackrefTree can contain a two different extent (LADDR_LEAF, BACKREF_LEAF) that have a same paddr (but, laddrs are different). This cause a problem when mounting seastore, because mark_space_used with address 1 is called twice during scan_mapped_space.

The solution in my mind is to call mark_space_free() to free paddr after remove_mapping() is called. What do you think?
or scan_mapped_space changes---apply alloc_infos, then iterate the tree---is better?

**Comment by cyx1231st** [comment]
> 1. Overwrite dirty extents: ... this means that we don't need to create a new extent, just overwriting an as-is extent to the original extent is enough, while increasing the version number ...

The current delta replay is not an idempotent procedure (e.g. tree node delta replay). So I agree that the issue is about how to make the dirty-extent inplace-rewrite atomic with the extent-version update.

> In terms of transaction B, after reusing address 1 due to duplication_for_write(), address 1 is replayed by new_mapping(). At this point, if trim_alloc is stopped, BackrefTree can contain a two different extent (LADDR_LEAF, BACKREF_LEAF) that have a same paddr (but, laddrs are different).

I don't follow why it will result in BackrefTree containing two different extents with the same paddr: transactions should enforce all-or-nothing in both memory and disk, it is either successful (BACKREF_LEAF is rewritten) or not (BACKREF_LEAF is not rewritten). This is also why in the transactions, `mark_space_free()` for its retired extents is called in the atomic `Cache::complete_commit()` after the transaction is successfully committed, not when the transaction is still pending.

btw, I'll take AL in the following two weeks, my responses might be slow.

**Comment by myoungwon** [comment]
Transaction sequence is: A -> B -> C -> D 

...
- Transaction A (LADDR_LEAF (paddr 1234) is allocated and alloc_info (paddr 1234) is stored)
- Transaction B (mutate, duplicate_for_write(LADDR_INTERNAL), existing LADDR_LEAF (paddr 1234) is retired) <-- paddr 1234 is free here
- Background Transaction C (trim_alloc,  duplicate_for_write(BACKREF_INTERNAL) by other changes, new BACKREF_LEAF is allocated (at this point, paddr 1234 is reused because it is free)) 
- Background Transaction D (trim_alloc, replays alloc_info(paddr 1234) according to Transaction A's alloc_info)
- Trim_alloc is stopped due to threshold (e.g., length)
- Remount starts

After transaction D, paddr 1234 is inserted twice.
Backreftree is not up-to-update even though alloc_info is persistent after transaction is done---the tree will be updated later after trim_alloc is called, right?
So, it seems that we need to reserve paddr 1234 or call new_mapping right after BACKREF_LEAF is allocated?

**Comment by myoungwon** [comment]
I think a problem as above is caused by a mismatch between the backref tree and alloc_delta_t, even though the changes on the backref tree are stored by alloc_delta_t---replaying alloc_delta is required to get up-to-date extents.

A solution to resolve this issue is that transaction includes changes regarding adding or removing or mutating btree node before writing extent using OOL, not using alloc_delta_t. Can you check the last commit I added?

**Comment by cyx1231st** [comment]
@myoungwon sorry for the late reply. I just back to work and I'm still trying to prioritize the pending reviews.

> * Transaction A (LADDR_LEAF (paddr 1234) is allocated and alloc_info (paddr 1234) is stored)
> * Transaction B (mutate, duplicate_for_write(LADDR_INTERNAL), existing LADDR_LEAF (paddr 1234) is retired) <-- paddr 1234 is free here
> * Background Transaction C (trim_alloc, duplicate_for_write(BACKREF_INTERNAL) by other changes, new BACKREF_LEAF is allocated (at this point, paddr 1234 is reused because it is free))
> * Background Transaction D (trim_alloc, replays alloc_info(paddr 1234) according to Transaction A's alloc_info)
> * Trim_alloc is stopped due to threshold (e.g., length)
> * Remount starts
> 
> After transaction D, paddr 1234 is inserted twice.
Backreftree is not up-to-update even though alloc_info is persistent after transaction is done---the tree will be updated later after trim_alloc is called, right?

I think you are right, and the issue only happens with RBM's `ExtentAllocator`.

From the current `JournalTrimmerImpl::trim_alloc()` logic, there is nothing to prevent the TRIM_ALLOC transactions from allocating an available paddr through `ExtentAllocator`, even though this paddr will conflict to a follow-up allocation attempt from an un-trimmed `backref_entry_t`.

**Comment by myoungwon** [comment]
> At Transaction C's commit time, paddr 1234 is reused but this won't apply to the backref tree directly. It will be appended to Cache::backref_entryrefs_by_seq instead, as backref_entry_t: 

That is true in case of segment. However, with random metadata (*_INTERNAL and *_LEAF) alllocation strategy (the pupose of this PR), new backref node can be directly stored to RBM.

**Comment by cyx1231st** [comment]
> That is true in case of segment.

My conclusion above was that the original trim-alloc logic should work with RBM without problem. If it is true, whether we should change the common trim-alloc behavior would be a different topic about optimization, not enabling, see below.

>  However, with random metadata (*_INTERNAL and *_LEAF) alllocation strategy (the goal of this PR), new backref node can be directly stored to RBM.
>
> Note that, with the last commit, we don't need trim_alloc() anymore.

IIUC the original intent of trim_alloc() is not related to segments. Like trim_dirty() is to merge updates to the persistent extents with `delta_info_t`, trim_alloc() is able to merge updates to the backref-tree with `backref_entry_t`. I think it is reasonable to further optimize `trim_alloc()` logic to resemble `trim_dirty()` for both performance and code-maintainence reasons instead of dropping it. (cc @xxhdx1985126 @athanatos )

So I'm not convinced with the following questions:
* Why is it related to the goal to enable writing physical extents as OOL? [1]
* As an optimization, what is the benefit over the existing trim_alloc() which is common and can benefit from the merged updates?

[1] https://github.com/ceph/ceph/blob/9754cafc029e1da83f5ddd4332b69066fe6b3ffb/src/crimson/os/seastore/extent_placement_manager.h#L223-L227

**Comment by myoungwon** [comment]
> My conclusion above was that the original trim-alloc logic should work with RBM without problem. If it is true, whether we should change the common trim-alloc behavior would be a different topic about optimization, not enabling, see below.

Without this PR changes, RBM works with existing trim-alloc.  This PR is optimization as you mentioned. Sorry for confusion.

> IIUC the original intent of trim_alloc() is not related to segments. Like trim_dirty() is to merge updates to the persistent extents with delta_info_t, trim_alloc() is able to merge updates to the backref-tree with backref_entry_t. I think it is reasonable to further optimize trim_alloc() logic to resemble trim_dirty() for both performance and code-maintainence reasons instead of dropping it. 

I didn't mean that trim_alloc() needs to be dropped---it is required for segment case. Only in RBM case, we don't need to call trim_alloc() with this PR changes---only trim_dirty() is required. 

> So I'm not convinced with the following questions:
>   Why is it related to the goal to enable writing physical extents as OOL? [1]

This is because a new metadata extent (backref, laddr) can be stored to RBM, I think the issue above is only related to RBM, not SegmentManager.

>    As an optimization, what is the benefit over the existing trim_alloc() which is common and can benefit from the merged updates?

We don't need consider two types of transactions: 1) trim_alloc and 2) trim_dirty. This make seastore's background work simple and predictable. Moreover, CBJournal contains smaller number of records than existing case, because there is only dirtied delta. This can reduce background work overhead---new extent is allocated to RBM, and delta is appended to Journal.

**Comment by cyx1231st** [comment]
> Without this PR changes, RBM works with existing trim-alloc. This PR is optimization as you mentioned. Sorry for confusion.
>
>> Why is it related to the goal to enable writing physical extents as OOL? [1]
>
> This is because a new metadata extent (backref, laddr) can be stored to RBM, I think the issue above is only related to RBM, not SegmentManager.
>
> In summary, I encountered the issue when I tried to implement [1]. ...

Probably I didn't make myself clear enough. My point in https://github.com/ceph/ceph/pull/49605#issuecomment-1418672654 is that RBM OOL physical extents enabling (the goal of this PR) should work with the existing trim-alloc in your example. So I think the proposal is independendent to the enabling as an optimization.

>> IIUC the original intent of trim_alloc() is not related to segments. Like trim_dirty() is to merge updates to the persistent extents with delta_info_t, trim_alloc() is able to merge updates to the backref-tree with backref_entry_t. I think it is reasonable to further optimize trim_alloc() logic to resemble trim_dirty() for both performance and code-maintainence reasons instead of dropping it.
>
> I didn't mean that trim_alloc() needs to be dropped---it is required for segment case. Only in RBM case, we don't need to call trim_alloc() with this PR changes---only trim_dirty() is required.
>
> We don't need consider two types of transactions: 1) trim_alloc and 2) trim_dirty. This make seastore's background work simple and predictable. Moreover, CBJournal contains smaller number of records than existing case, because there is only dirtied delta. This can reduce background work overhead and WAF---new extent is allocated to RBM, and delta is appended to Journal.

As an optimization, I'm trying to understand why trim_alloc() should be dropped (and specialized) in the RBM case. For both segments and RBM, the existing trim_alloc() is able to merge mutiple updates (which are more expensive with regard to both overhead and WAF) to the backref-tree by appending `backref_entry_t` instead.

**Comment by myoungwon** [comment]
Hmm, without other changes (partial overwrite), let me clarify this PR.
The goal of this PR is to implement [1].
But, I encountered the problem above. So, I suggested the following code in the last commit to resolve the issue.

```
    return seastar::do_with(
      tref.get_retired_set(),
      [this, FNAME, &tref] (auto &extents) {
      return trans_intr::do_for_each(
	extents,
	[this, &tref] (CachedExtentRef extent) {
	if (!epm->is_random_allocation()) {
	  return submit_transaction_iertr::now();
	}
	return backref_manager->remove_mapping(
	  tref,
	  extent->get_paddr()
	).si_then([](auto&& ret) {
	  return seastar::now();
	}).handle_error_interruptible(
	  crimson::ct_error::input_output_error::pass_further(),
	  crimson::ct_error::assert_all("no enoent possible")
	);
      }).si_then([this, FNAME, &tref] {
	return seastar::do_with(
	  tref.get_valid_alloc_info_list(epm->is_random_allocation()),
	  [this, FNAME, &tref] (auto &allocated_extents) {
	  return trans_intr::do_for_each(
	    allocated_extents,
	    [this, &tref] (CachedExtentRef extent) {
	    laddr_t laddr;
	    if (extent->is_logical()) {
	      laddr = extent->cast<LogicalCachedExtent>()->get_laddr();
	    } else if (is_lba_node(extent->get_type())) {
	      laddr = extent->cast<lba_manager::btree::LBANode>()->get_node_meta().begin;
	    }
	    assert(!is_backref_node(extent->get_type()));
	    return backref_manager->new_mapping(
	      tref,
	      extent->get_paddr(),
	      extent->get_length(),
	      laddr,
	      extent->get_type()
	    ).si_then([](auto &&pin) {
	      return seastar::now();
	    });
	  }).si_then([this, FNAME, &tref] {
	    auto allocated_extents = tref.get_valid_pre_alloc_list();
	    auto num_extents = allocated_extents.size();
	    SUBTRACET(seastore_t, "process {} allocated extents", tref, num_extents);
	    return epm->write_preallocated_ool_extents(tref, allocated_extents
	    ).si_then([] {
	      return seastar::now();
	    });
	  });
	});
      });
    });
```

I think that bypassing trim_alloc() is incidental detail. We need to discuss how to overcome the issue first.


>> Background Transaction C (trim_alloc, duplicate_for_write(BACKREF_INTERNAL) by other changes, new BACKREF_LEAF is allocated (at this point, paddr 1234 is reused because it is free))

> At Transaction C's commit time, paddr 1234 is reused but this won't apply to the backref tree directly. It will be appended to Cache::backref_entryrefs_by_seq instead, as backref_entry_t:

At transaction C (trim alloc) 's commit time, BACKREF_LEAF can be allocated at 1234 in RBM, not journal (this is the goal of this PR)---trim_alloc allocate BACKREF_LEAF extent, referring alloc_info, right?.

**Comment by cyx1231st** [comment]
> We need to focus on how to overcome the issue first. Hmm, how can we avoid this with keeping the existing mechanism?

Yeah, I still need to understand why there is a problem in the existing implementation first, before going forward to discuss the solution.

The current trim-alloc is there because of the state of the persistent backref-tree is **intentially** not up-to-date, and by applying alloc-deltas starting from the alloc-tail, the backref-tree can be consistent with the latest allocations (i.e. the current ExtentAllocator state in the RBM case), for example:
```
[TransA] (alloc-tail) [TransB] [TransC] [TransD] (journal-head) =>
     ^                                      ^                                           
     |                                      |
backref-tree + alloc-deltas(A,B,C,D) = ExtentAllocator
```

In other words, the backref-tree will never represent (shouldn't interfere with) the latest allocations (i.e. ExtentAllocator). And the tree is updated independently in the trim-alloc background task.

Back to your example https://github.com/ceph/ceph/pull/49605#issuecomment-1386366820:

> * Transaction A (LADDR_LEAF (paddr 1234) is allocated and alloc_info (paddr 1234) is stored)
> * Transaction B (mutate, duplicate_for_write(LADDR_INTERNAL), existing LADDR_LEAF (paddr 1234) is retired) <-- paddr 1234 is free here
> * Background Transaction C (trim_alloc, duplicate_for_write(BACKREF_INTERNAL) by other changes, new BACKREF_LEAF is allocated (at this point, paddr 1234 is reused because it is free))
> * Background Transaction D (trim_alloc, replays alloc_info(paddr 1234) according to Transaction A's alloc_info)
>
> Trim_alloc is stopped due to threshold (e.g., length)

I don't really see there is a problem:
* At Transaction C's commit time, the on-going allocation (alloc paddr 1234) is happened in the `ExtentAllocator`.
* At Transaction D's commit time, the `backref-tree` has inserted paddr 1234 according to A's alloc-delta.

The both updates should be independent to each other and won't conflict.

> Remount starts
> After transaction D, paddr 1234 is inserted twice.

At replay time (remount), the backref tree will apply alloc-deltas sequentially starting from the alloc-tail.

In your example, since Trans D is already committed, the alloc-tail is at Trans A, and the replay will start from Trans B. This means that the backref-tree will apply B's alloc-delta first before applying C's alloc-delta, so paddr 1234 will be erased first (B) then inserted (C) in the tree. So I don't see there is an "insert-twice" problem.

**Comment by myoungwon** [comment]
>I don't really see there is a problem:
> At Transaction C's commit time, the on-going allocation (alloc paddr 1234) is happened in the ExtentAllocator.
> At Transaction D's commit time, the backref-tree has inserted paddr 1234 according to A's alloc-delta.

Transaction C is trim_alloc()---note that alloc-delta in Transaction C is recorded before Transaction A. trim_alloc() updates backref tree, am I missing something? If right, the trim_alloc() generates a new BACKREF_LEAF extent and appends it to the journal---this is an existing implementation except for this PR and it works. 
But, with [1] the BACKREF_LEAF extent can be directly stored to the RBM using OOL path, so the BACKREF_LEAF extent itself occupies paddr 1234 in RBM, not journal. As a result, after transaction C, backref-tree contains paddr 1234 (because it is used by the BACKREF_LEAF).

Then, after transaction D, the backref-tree has inserted paddr 1234 according to A's alloc-delta.

**Comment by cyx1231st** [comment]
> Transaction C is trim_alloc()---note that alloc-delta in Transaction C is recorded before Transaction A. trim_alloc() updates backref tree, am I missing something? If right, the trim_alloc() generates a new BACKREF_LEAF extent and appends it to the journal---this is an existing implementation except for this PR and it works.

Right, before C applies alloc-delta from A (alloc paddr 1234), it is possible for a new BACKREF_LEAF to be at the paddr 1234, as it is available from the ExtentAllocator. 

> But, to implement [1], the BACKREF_LEAF extent needs to be directly stored to the RBM using OOL path, so the BACKREF_LEAF extent itself occupies paddr 1234 in RBM after OOL write, not journal.

Yeah, BACKREF_LEAF occupies paddr 1234 (as an entry value in a BACKREF_INTERNAL node), but IIUC it is independent to insert paddr 1234 into the backref-tree (as an entry key in a BACKREF_LEAF node) at the same time.

> As a result, after transaction C, backref-tree contains paddr 1234 (because it is used by the BACKREF_LEAF extent and backref tree should keep track of block allocation in RBM).

"contains" here seems inaccurate to me, see the differences above.

> Then, after transaction D, the backref-tree has inserted paddr 1234 according to A's alloc-delta.

I think it is independent (there is no problem) to insert an entry key 1234 in a BACKREF_LEAF node, and insert an entry value 1234 in a BACKREF_INTERNAL node.

**Comment by myoungwon** [comment]
Hmm, at starup time (remount), isn't mark_space_free(paddr 1234) called twice, 1) BACKREF_LEAF  and 2) LADDR_LEAF, during scan_mapped_space() before handling backref_entry_mset in scan_mapped_space()?

**Comment by cyx1231st** [comment]
> Hmm, at starup time (remount), isn't mark_space_free(paddr 1234) called twice, 1) BACKREF_LEAF and 2) LADDR_LEAF, during scan_mapped_space() before handling backref_entry_mset in scan_mapped_space()?

I mark_space_* are be called after journal replay, and they should in the following sequence according to `BtreeBackrefManager::scan_mapped_space()`:
* mark_space_used(paddr 1234): from backref-tree traversal as D is committed
* mark_space_free(paddr 1234): from the loaded alloc-delta of B
* mark_space_used(paddr 1234): from the loaded alloc-delta of C

That should be fine, right?

**Comment by myoungwon** [comment]
> mark_space_* are called after journal replay, and they should be in the following sequence according to BtreeBackrefManager::scan_mapped_space():

 >   mark_space_used(paddr 1234): from backref-tree traversal as D is committed
    mark_space_free(paddr 1234): from the loaded alloc-delta of B
    mark_space_used(paddr 1234): from the loaded alloc-delta of C

> That should be fine, right?

That sequence is fine, but the sequence above example is:

mark_space_used (paddr 1234): C's BACKREF_LEAF  from backref-tree traversal 
mark_space_used (paddr 1234): D's LADDR_LEAF  from backref-tree traversal 
mark_space_free(paddr 1234): from the loaded alloc-delta of B

right?

**Comment by cyx1231st** [comment]
> That sequence is fine, but the sequence above example is:

> mark_space_used (paddr 1234): C's BACKREF_LEAF from backref-tree traversal
> mark_space_used (paddr 1234): D's LADDR_LEAF from backref-tree traversal
> mark_space_free(paddr 1234): from the loaded alloc-delta of B

> Am I wrong?

> We might avoid this if we skip mark_space_used(paddr 1234) caused by D by checking that there is deleted alloc-delta.
But, I think this is a workaround.

Ah, thanks. I think I understand where is missing -- the allocations of backref leaf/internal nodes represent the latest (ExtentAllocator) state, not the alloc-tail state. This is why we will see conflicts if the apply order is wrong.

A more reasonable fix to this issue might be to correct the order by postponing the backref internal-node entry traversal.

So the high-level logic in `BtreeBackrefManager::scan_mapped_space()` would become:
1. Traverse the allocations from backref-tree leaf-node entries;
2. Traverse the pending alloc-deltas by sequence;
3. Traverse the allocations from backref-tree internal-node entries;

Does this also make sense to you @xxhdx1985126 ?

**Comment by myoungwon** [comment]
OK, if @xxhdx1985126 agrees on this, I'll make changes. Then, post partial overwrite PR (this is not related with trim_alloc). Make sense?

**Comment by cyx1231st** [comment]
Please go ahead if the fix is reasonable to you. I will mention it in the next standup.

> Then, post partial overwrite PR (this is not related with trim_alloc).

Hmm, I suppose this is another feature that is not related to `enable metadata (*_INTERNAL and *_LEAF ) allocation using OOL path` ?

**Comment by myoungwon** [comment]
> Hmm, I suppose this is another feature that is not related to enable metadata (*_INTERNAL and *_LEAF ) allocation using OOL path ?

Right. I meant that I'll post it as a separated PR.

**Comment by myoungwon** [comment]
@cyx1231st Can you take a look?

File: src/crimson/os/seastore/extent_placement_manager.h [hunk 1]
```diff
@@ -220,7 +230,7 @@ class ExtentPlacementManager {
 	      gen};
     };
 
-    if (!is_logical_type(type)) {
+    if (should_inline(type)) {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/extent_placement_manager.h:233]: IIUC the situation after this PR is that all extents with RBM must be OOL.

If true, probably we can adjust to the following logic with the explict restrictions:
```c++
if (type == extent_types_t::ROOT) {
  return alloc_paddr(INLINE_GENERATION, data_category_t::METADATA, length);
}

if (background_process.get_backend_type() == backend_type_t::SEGMENTED &&
    is_lba_backref_node(type)) {
  // with SEGMENTED, lba-backref extents must be INLINE
  return alloc_paddr(INLINE_GENERATION, data_category_t::METADATA, length);
}

if (hint == placement_hint_t::COLD) {
  assert(gen == INIT_GENERATION);
  return alloc_paddr(MIN_REWRITE_GENERATION, get_extent_category(type), length);
}

if (get_extent_category(type) == data_category_t::METADATA &&
    gen == INIT_GENERATION) {
  if (background_process.get_backend_type() == backend_type_t::SEGMENTED) {
    // with SEGMENTED, default not to ool metadata extents to reduce padding overhead.
    // TODO: improve padding so we can default to the ool path.
    return alloc_paddr(INLINE_GENERATION, get_extent_category(type), length);
  } else {
    // with RBM, all extents must be OOL
    assert(background_process.get_backend_type() == backend_type_t::RANDOM_BLOCK);
    return alloc_paddr(OOL_GENERATION, get_extent_category(type), length);
  }
} else {
  assert(get_extent_category(type) == data_category_t::DATA ||
         gen >= MIN_REWRITE_GENERATION);
  if (gen > MAX_REWRITE_GENERATION) {
    gen = MAX_REWRITE_GENERATION;
  } else if (gen == INIT_GENERATION) {
    gen = OOL_GENERATION;
  }
  return alloc_paddr(gen, get_extent_category(type), length);
}
```

Related adjustments:
* Drop the deprecated prefer_ool.
* Introduce BackgroundProcess::get_backend_type(), should be available from trimmer.
* Introduce is_lba_backref_node() in seastore_types.h

File: src/crimson/os/seastore/btree/fixed_kv_btree.h [hunk 2]
```diff
@@ -836,9 +836,13 @@ class FixedKVBtree {
        * Upon commit, these now block relative addresses will be interpretted
```
Comment (cyx1231st) [inline src/crimson/os/seastore/btree/fixed_kv_btree.h:836]: Switch to the newly introduced is_lba_backref_node() at 809 above.

**Review by cyx1231st — COMMENTED** [review]
The change looks good!

My suggestions are only about readability improvements.

**Comment by xxhdx1985126** [comment]
> So the high-level logic in `BtreeBackrefManager::scan_mapped_space()` would become:
> 
> 1. Traverse the allocations from backref-tree leaf-node entries;
> 2. Traverse the pending alloc-deltas by sequence;
> 3. Traverse the allocations from backref-tree internal-node entries;
> 
> Does this also make sense to you @xxhdx1985126 ?

Sorry for the late reply, I'm still trying to catch up with the discussions, but, being specific to this strategy of scanning the backref tree, I think it's fine.:-)

**Comment by myoungwon** [comment]
@cyx1231st I addressed your comments. Please take a look at this.

File: src/crimson/os/seastore/async_cleaner.cc [hunk 3]
```diff
@@ -1752,9 +1752,21 @@ bool RBMCleaner::equals(const RBMSpaceTracker &_other) const
...
+	if (p.paddr == paddr) {
+	  match = true;
+	  break;
+	}
+      }
+      if (!match) {
+	all_match = false;
+	ERROR("block addr {} mismatch other used: {}",
+	      addr, i->second.used);
+      }
```
Comment (cyx1231st) [inline src/crimson/os/seastore/async_cleaner.cc:1769]: Why is this necessary?

IIUC this logic implies that the rbm allocator may not contain the allocations of the backref-tree nodes.

But it doesn't match my understanding that `epm.commit/mark_space_used` are including backref-tree nodes in both transaction commit and replay paths, so the rbm allocator should contain the allocations from the backref-tree nodes.

Or am I still missing anything?

File: src/crimson/os/seastore/extent_placement_manager.h (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/extent_placement_manager.h]: cc @zhscn

**Review by cyx1231st — COMMENTED** [review]
I think this is close except for my question, @athanatos do you want to take another look?

File: src/crimson/os/seastore/async_cleaner.cc (hunk 3 above)
Comment (myoungwon) [reply src/crimson/os/seastore/async_cleaner.cc:1769]: `other`is the result (RBMSpaceTracker) from `scan_mapped_space`---therefore, it contains current backref tree's information about block allocation. At this point, ExtentAllocator(RBM) keeps track of the allocations, but the backref tree's result does not because alloc-deltas are not yet replayed. So, this part checks if the address is allocated for backref node. Am I missing?

Comment (myoungwon) [reply src/crimson/os/seastore/async_cleaner.cc]: Sorry, I was mistaken about error log I captured before. Removed.

**Review by cyx1231st — APPROVED** [review]
